### PR TITLE
Add support for PaymentMethod

### PR DESCRIPTION
**Pull request by remi-stripe**
cc @stripe/api-libraries

### CODE REVIEW

File: charge.go [hunk 1]
```diff
@@ -144,6 +167,194 @@ type ChargeLevel3 struct {
...
+}
+
+// ChargePaymentMethodDetailsCardThreeDSecure represents details about 3DS associated with the
+// charge's PaymentMethod.
+type ChargePaymentMethodDetailsCardThreeDSecure struct {
+}
+
+// PaymentMethodDetailsCardWallet represents the details of the card wallet if this Card
+// PaymentMethod is part of a card wallet.
+type PaymentMethodDetailsCardWallet struct {
```
Comment (mickjermsurawong-stripe) [inline charge.go]: @remi-stripe  should this be prefixed with `ChargeXXX` like other structs too?

Comment (remi-stripe) [reply charge.go]: Yes it should, nice catch.

File: paymentmethod.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	PaymentMethodCardWalletTypeGooglePay           PaymentMethodCardWalletType = "google_pay"
+	PaymentMethodCardWalletTypeMasterpass          PaymentMethodCardWalletType = "masterpass"
+	PaymentMethodCardWalletTypeSamsungPay          PaymentMethodCardWalletType = "samsung_pay"
+	PaymentMethodCardWalletTypeVisaCheckout        PaymentMethodCardWalletType = "visa_checkout"
+)
+
+// PaymentMethodTransferDataParams is the set of parameters allowed for the transfer_data hash.
+type PaymentMethodTransferDataParams struct {
+	Destination *string `form:"destination"`
+}
```
Comment (garrett-stripe) [inline paymentmethod.go]: I think this is unintentional—this struct doesn't appear to be used anywhere.

Comment (remi-stripe) [reply paymentmethod.go]: ah yes bad copy paste, fixed (won't push while you review)

File: paymentintent.go [hunk 3]
```diff
@@ -114,14 +114,16 @@ type PaymentIntentListParams struct {
 
 // PaymentIntentLastPaymentError represents the last error happening on a payment intent.
 type PaymentIntentLastPaymentError struct {
```
Comment (remi-stripe) [inline paymentintent.go]: Technically, this should be `Error` right? Is it worth changing this and releasing as a major version to be cleaner?

**Comment by remi-stripe** [comment]
Okay, this should now be feature complete and ready for review.

There is one test failing, `TestTokenNew_SharedCustomerBankAccount` and it's because of a change in openapi which does not expect `bank_account` to be a string even though it can be. cc @mickjermsurawong-stripe as this is likely due to a change in our codebase for stripe-java autogeneration.

r? @brandur-stripe

**Comment by remi-stripe** [comment]
@mickjermsurawong-stripe Thanks a lot for the thorough review. I can't believe after all the back and forth I did on those PRs I still missed the wallet fields on Charge. I fixed everything you found I think. PTAL!

**Comment by mickjermsurawong-stripe** [comment]
thanks remi for the fix!
LGTM

**Comment by mickjermsurawong-stripe** [comment]
ah the test is failing on `/v1/tokens` as it is trying to create bank account token with a token. 
My change on typed param work requires only the bank_account object instead of hash. 
But the failing test here is a valid use case we want to support? I can add this back to support string token as well. @remi-stripe 

```
func TestTokenNew_SharedCustomerBankAccount(t *testing.T) {
	params := &stripe.TokenParams{
		BankAccount: &stripe.BankAccountParams{
			ID: stripe.String("ba_123"),
		},
		Customer: stripe.String("cus_123"),
	}
	params.SetStripeAccount("acct_123")
	token, err := New(params)
	assert.Nil(t, err)
	assert.NotNil(t, token)
}
```

**Comment by remi-stripe** [comment]
You can ignore that failing test, it's being fixed in another PR in parallel I just didn't want to duplicate the work

File: paymentmethod.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+// List of values that PaymentMethodCardBrand can take.
+const (
+	PaymentMethodCardBrandAmex       PaymentMethodCardBrand = "amex"
+	PaymentMethodCardBrandDiners     PaymentMethodCardBrand = "diners"
+	PaymentMethodCardBrandDiscover   PaymentMethodCardBrand = "discover"
+	PaymentMethodCardBrandJCB        PaymentMethodCardBrand = "jcb"
+	PaymentMethodCardBrandMastercard PaymentMethodCardBrand = "mastercard"
+	PaymentMethodCardBrandUnionpay   PaymentMethodCardBrand = "unionpay"
+	PaymentMethodCardBrandVisa       PaymentMethodCardBrand = "visa"
+	PaymentMethodCardBrandUnknown    PaymentMethodCardBrand = "unknown"
```
Comment (brandur-stripe) [inline paymentmethod.go]: Minor alphabetization problem here: unknown should be above visa.

Comment (remi-stripe) [reply paymentmethod.go]: Nice catch, fixed

**Comment by remi-stripe** [comment]
I had to remove a test because we don't support `bank_account: "ba_123"` when creating tokens anymore. cc @mickjermsurawong-stripe

**Comment by mickjermsurawong-stripe** [comment]
LGTM

**Pull request merged**